kubernetes_ingress_v1 resource does not correctly include empty tls configuation blocks in resulting k8s ingress object

<!---
Hi there,

Thank you for opening an issue. Please note that we try to keep the Terraform issue tracker reserved for bug reports and feature requests. For general usage questions, please see: https://www.terraform.io/community.html.
--->


### Terraform Version, Provider Version and Kubernetes Version
<!--- Run `terraform -v` to show the version. If you are not running the latest version of Terraform, please upgrade because your issue may have already been fixed. --->
```
Terraform version: v1.6.2
Kubernetes provider version: v2.23.0
Kubernetes version: v1.26.9
```

### Affected Resource(s)
<!-- Please list the resources as a list, for example:
- opc_instance
- opc_storage_volume
If this issue appears to affect multiple resources, it may be an issue with Terraform's core, so please mention this. -->
kubernetes_ingress_v1. No others tested

### Terraform Configuration Files
```hcl
resource "kubernetes_ingress_v1" "myingress" {
  metadata {
    name      = "myingress"
  }
  spec {
    default_backend {
      service {
        name = "myservice"
        port {
          number = 80
        }
      }
    }
    tls {}
  }
}
```

### Steps to Reproduce
1. `terraform apply`
2. `kubectl get ing myingress -o yaml`

### Expected Behavior
Ingress should have an empty tls block

### Actual Behavior
Ingress has no tls blocks

### Important Factoids
This behavior is important to support because an ingress controller may treat an empty tls config different from a missing tls config, like in OpenShift. Also something to note: multiple empty tls blocks cause a panic.


### Community Note
<!--- Please keep this note for the community --->
* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

The behavior I described is documented here: https://docs.openshift.com/container-platform/4.14/networking/routes/route-configuration.html#creating-edge-route-with-default-certificate_route-configuration

```
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: frontend
  ...
spec:
  rules:
    ...
  tls:
  - {}
```

> 	Use this exact syntax to specify TLS without specifying a custom certificate.

The panic I referred to looks like this:
```
panic: interface conversion: interface {} is nil, not map[string]interface {}

goroutine 445 [running]:
github.com/hashicorp/terraform-provider-kubernetes/kubernetes.expandIngressV1TLS({0xc0026ddf00, 0x2, 0x33d9cc6?})
        /root/terraform-provider-kubernetes/kubernetes/structure_ingress_spec_v1.go:261 +0x2cd
github.com/hashicorp/terraform-provider-kubernetes/kubernetes.expandIngressV1Spec({0xc002732380?, 0x33da3fc?, 0x0?})
        /root/terraform-provider-kubernetes/kubernetes/structure_ingress_spec_v1.go:200 +0x265
github.com/hashicorp/terraform-provider-kubernetes/kubernetes.resourceKubernetesIngressV1Create({0x380bef8?, 0xc001b22770}, 0xc0019f9780, {0x32f0240?, 0xc00134fe30})
        /root/terraform-provider-kubernetes/kubernetes/resource_kubernetes_ingress_v1.go:184 +0x1cf
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).create(0xc000f688c0, {0x380be50, 0xc002962f90}, 0xd?, {0x32f0240, 0xc00134fe30})
        /root/go-workspace/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.24.1/helper/schema/resource.go:707 +0x11b
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0xc000f688c0, {0x380be50, 0xc002962f90}, 0xc001110b60, 0xc00293fa00, {0x32f0240, 0xc00134fe30})        /root/go-workspace/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.24.1/helper/schema/resource.go:837 +0xa69
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0xc0014f0168, {0x380be50?, 0xc002962f00?}, 0xc00213f4a0)
        /root/go-workspace/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.24.1/helper/schema/grpc_provider.go:1021 +0xdbc
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ApplyResourceChange(0xc000334a00, {0x380be50?, 0xc0029626f0?}, 0xc00293a460)
        /root/go-workspace/pkg/mod/github.com/hashicorp/terraform-plugin-go@v0.14.2/tfprotov5/tf5server/server.go:818 +0x56a
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ApplyResourceChange_Handler({0x32f42e0?, 0xc000334a00}, {0x380be50, 0xc0029626f0}, 0xc00293a3f0, 0x0)
        /root/go-workspace/pkg/mod/github.com/hashicorp/terraform-plugin-go@v0.14.2/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:385 +0x169
google.golang.org/grpc.(*Server).processUnaryRPC(0xc00101c3c0, {0x3815300, 0xc001cc69c0}, 0xc002957e60, 0xc002294570, 0x4bc0f80, 0x0)
        /root/go-workspace/pkg/mod/google.golang.org/grpc@v1.56.3/server.go:1335 +0xde7
google.golang.org/grpc.(*Server).handleStream(0xc00101c3c0, {0x3815300, 0xc001cc69c0}, 0xc002957e60, 0x0)
        /root/go-workspace/pkg/mod/google.golang.org/grpc@v1.56.3/server.go:1712 +0x9e7
google.golang.org/grpc.(*Server).serveStreams.func1.1()
        /root/go-workspace/pkg/mod/google.golang.org/grpc@v1.56.3/server.go:947 +0xbb
created by google.golang.org/grpc.(*Server).serveStreams.func1 in goroutine 423
        /root/go-workspace/pkg/mod/google.golang.org/grpc@v1.56.3/server.go:958 +0x145
FAIL    github.com/hashicorp/terraform-provider-kubernetes/kubernetes   5.618s
FAIL
make: *** [GNUmakefile:83: testacc] Error 1
```

It occurs when you use multiple TLS blocks. The first must be non-empty and a later one must be empty. This is because the `expandIngressV1TLS` function does not handle multiple blocks (or empty blocks) properly.

```
func expandIngressV1TLS(l []interface{}) []networking.IngressTLS {
  if len(l) == 0 || l[0] == nil {
    return nil
  }

  tlsList := make([]networking.IngressTLS, len(l))
  for i, t := range l {
    in := t.(map[string]interface{})
    obj := networking.IngressTLS{}

    if v, ok := in["hosts"]; ok {
      obj.Hosts = expandStringSlice(v.([]interface{}))
    }

    if v, ok := in["secret_name"].(string); ok {
      obj.SecretName = v
    }
    tlsList[i] = obj
  }

  return tlsList
}
```

The code checks if the first block is `nil`, and then, if not, assumes none of the remaining blocks are `nil`. If this assumption is false the code panics.

--- a/.changelog/2344.txt
+++ b/.changelog/2344.txt
@@ -1,0 +1,7 @@
+```release-note:bug
+`resource/kubernetes_ingress`: Fix an issue where the empty `tls` attribute in the configuration does not generate the corresponding Ingress object without any TLS configuration.
+```
+
+```release-note:bug
+`resource/kubernetes_ingress_v1`: Fix an issue where the empty `tls` attribute in the configuration does not generate the corresponding Ingress object without any TLS configuration.
+```

--- a/kubernetes/provider_ignore_metadata_test.go
+++ b/kubernetes/provider_ignore_metadata_test.go
@@ -44,8 +44,7 @@ func TestAccKubernetesIgnoreKubernetesMetadata_basic(t *testing.T) {
 }
 
 func testAccKubernetesIgnoreKubernetesMetadataProviderConfig(namespaceName string, ignoreKubernetesMetadata string) string {
-	return fmt.Sprintf(`
-provider "kubernetes" {
+	return fmt.Sprintf(`provider "kubernetes" {
   ignore_annotations = [
     "%s",
   ]

--- a/kubernetes/resource_kubernetes_ingress_v1_test.go
+++ b/kubernetes/resource_kubernetes_ingress_v1_test.go
@@ -164,6 +164,39 @@ func TestAccKubernetesIngressV1_TLS(t *testing.T) {
 	})
 }
 
+func TestAccKubernetesIngressV1_emptyTLS(t *testing.T) {
+	var conf networking.Ingress
+	name := fmt.Sprintf("tf-acc-test-%s", acctest.RandStringFromCharSet(10, acctest.CharSetAlphaNum))
+	resourceName := "kubernetes_ingress_v1.test"
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck: func() {
+			testAccPreCheck(t)
+			skipIfClusterVersionLessThan(t, "1.22.0")
+		},
+		IDRefreshName:     resourceName,
+		ProviderFactories: testAccProviderFactories,
+		CheckDestroy:      testAccCheckKubernetesIngressV1Destroy,
+		IDRefreshIgnore:   []string{"metadata.0.resource_version"},
+		Steps: []resource.TestStep{
+			{
+				Config: testAccKubernetesIngressV1Config_emptyTLS(name),
+				Check: resource.ComposeAggregateTestCheckFunc(
+					testAccCheckKubernetesIngressV1Exists(resourceName, &conf),
+					resource.TestCheckResourceAttr(resourceName, "metadata.0.name", name),
+					resource.TestCheckResourceAttrSet(resourceName, "metadata.0.generation"),
+					resource.TestCheckResourceAttrSet(resourceName, "metadata.0.resource_version"),
+					resource.TestCheckResourceAttrSet(resourceName, "metadata.0.uid"),
+					resource.TestCheckResourceAttr(resourceName, "spec.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "spec.0.tls.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "spec.0.tls.0.hosts.#", "0"),
+					resource.TestCheckResourceAttr(resourceName, "spec.0.tls.0.secret_name", ""),
+				),
+			},
+		},
+	})
+}
+
 func TestAccKubernetesIngressV1_InternalKey(t *testing.T) {
 	var conf networking.Ingress
 	name := fmt.Sprintf("tf-acc-test-%s", acctest.RandStringFromCharSet(10, acctest.CharSetAlphaNum))
@@ -536,6 +569,26 @@ func testAccKubernetesIngressV1Config_TLS_modified(name string) string {
     tls {
       hosts       = ["host1", "host2"]
       secret_name = "super-sekret"
+    }
+  }
+}`, name)
+}
+
+func testAccKubernetesIngressV1Config_emptyTLS(name string) string {
+	return fmt.Sprintf(`resource "kubernetes_ingress_v1" "test" {
+  metadata {
+    name = "%s"
+  }
+  spec {
+    default_backend {
+      service {
+        name = "app1"
+        port {
+          number = 443
+        }
+      }
+    }
+    tls {
     }
   }
 }`, name)

--- a/kubernetes/resource_kubernetes_ingress_v1beta1_test.go
+++ b/kubernetes/resource_kubernetes_ingress_v1beta1_test.go
@@ -119,6 +119,39 @@ func TestAccKubernetesIngressV1Beta1_TLS(t *testing.T) {
 	})
 }
 
+func TestAccKubernetesIngressV1Beta1_emptyTLS(t *testing.T) {
+	var conf api.Ingress
+	name := fmt.Sprintf("tf-acc-test-%s", acctest.RandStringFromCharSet(10, acctest.CharSetAlphaNum))
+	resourceName := "kubernetes_ingress.test"
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck: func() {
+			testAccPreCheck(t)
+			skipIfClusterVersionGreaterThanOrEqual(t, "1.22.0")
+		},
+		IDRefreshName:     resourceName,
+		IDRefreshIgnore:   []string{"metadata.0.resource_version"},
+		ProviderFactories: testAccProviderFactories,
+		CheckDestroy:      testAccCheckKubernetesIngressV1Beta1Destroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccKubernetesIngressV1Beta1Config_TLS(name),
+				Check: resource.ComposeAggregateTestCheckFunc(
+					testAccCheckKubernetesIngressV1Beta1Exists(resourceName, &conf),
+					resource.TestCheckResourceAttr(resourceName, "metadata.0.name", name),
+					resource.TestCheckResourceAttrSet(resourceName, "metadata.0.generation"),
+					resource.TestCheckResourceAttrSet(resourceName, "metadata.0.resource_version"),
+					resource.TestCheckResourceAttrSet(resourceName, "metadata.0.uid"),
+					resource.TestCheckResourceAttr(resourceName, "spec.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "spec.0.tls.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "spec.0.tls.0.hosts.#", "0"),
+					resource.TestCheckResourceAttr(resourceName, "spec.0.tls.0.secret_name", ""),
+				),
+			},
+		},
+	})
+}
+
 func TestAccKubernetesIngressV1Beta1_InternalKey(t *testing.T) {
 	var conf api.Ingress
 	name := fmt.Sprintf("tf-acc-test-%s", acctest.RandStringFromCharSet(10, acctest.CharSetAlphaNum))
@@ -194,7 +227,6 @@ func TestAccKubernetesIngressV1Beta1_WaitForLoadBalancerGoogleCloud(t *testing.T
 
 func testAccCheckKubernetesIngressV1Beta1Destroy(s *terraform.State) error {
 	conn, err := testAccProvider.Meta().(KubeClientsets).MainClientset()
-
 	if err != nil {
 		return err
 	}
@@ -304,6 +336,22 @@ func testAccKubernetesIngressV1Beta1Config_TLS(name string) string {
     tls {
       hosts       = ["host1"]
       secret_name = "super-sekret"
+    }
+  }
+}`, name)
+}
+
+func testAccKubernetesIngressV1Beta1Config_emptyTLS(name string) string {
+	return fmt.Sprintf(`resource "kubernetes_ingress" "test" {
+  metadata {
+    name = "%s"
+  }
+  spec {
+    backend {
+      service_name = "app1"
+      service_port = 443
+    }
+    tls {
     }
   }
 }`, name)

--- a/kubernetes/resource_kubernetes_mutating_webhook_configuration_test.go
+++ b/kubernetes/resource_kubernetes_mutating_webhook_configuration_test.go
@@ -216,8 +216,7 @@ func testAccCheckKubernetesMutatingWebhookConfigurationExists(n string) resource
 }
 
 func testAccKubernetesMutatingWebhookConfigurationConfig_basic(name string) string {
-	return fmt.Sprintf(`
-resource "kubernetes_mutating_webhook_configuration" "test" {
+	return fmt.Sprintf(`resource "kubernetes_mutating_webhook_configuration" "test" {
   metadata {
     name = %q
   }
@@ -254,8 +253,7 @@ resource "kubernetes_mutating_webhook_configuration" "test" {
 }
 
 func testAccKubernetesMutatingWebhookConfigurationConfig_modified(name string) string {
-	return fmt.Sprintf(`
-resource "kubernetes_mutating_webhook_configuration" "test" {
+	return fmt.Sprintf(`resource "kubernetes_mutating_webhook_configuration" "test" {
   metadata {
     name = %q
   }
@@ -306,8 +304,7 @@ resource "kubernetes_mutating_webhook_configuration" "test" {
 }
 
 func testAccKubernetesMutatingWebhookConfigurationConfig_without_rules(name string) string {
-	return fmt.Sprintf(`
-resource "kubernetes_mutating_webhook_configuration" "test" {
+	return fmt.Sprintf(`resource "kubernetes_mutating_webhook_configuration" "test" {
   metadata {
     name = %q
   }

--- a/kubernetes/resource_kubernetes_node_taint_test.go
+++ b/kubernetes/resource_kubernetes_node_taint_test.go
@@ -155,8 +155,7 @@ func testAccKubernetesNodeTaintExists(n string) resource.TestCheckFunc {
 }
 
 func testAccKubernetesNodeTaintConfig_basic() string {
-	return fmt.Sprintf(`
-data "kubernetes_nodes" "test" {}
+	return fmt.Sprintf(`data "kubernetes_nodes" "test" {}
 
 resource "kubernetes_node_taint" "test" {
   metadata {
@@ -173,8 +172,7 @@ resource "kubernetes_node_taint" "test" {
 }
 
 func testAccKubernetesNodeTaintConfig_multipleBasic() string {
-	return fmt.Sprintf(`
-data "kubernetes_nodes" "test" {}
+	return fmt.Sprintf(`data "kubernetes_nodes" "test" {}
 
 resource "kubernetes_node_taint" "test" {
   metadata {
@@ -201,8 +199,7 @@ resource "kubernetes_node_taint" "test" {
 }
 
 func testAccKubernetesNodeTaintConfig_updateTaint() string {
-	return fmt.Sprintf(`
-data "kubernetes_nodes" "test" {}
+	return fmt.Sprintf(`data "kubernetes_nodes" "test" {}
 
 resource "kubernetes_node_taint" "test" {
   metadata {
@@ -229,8 +226,7 @@ resource "kubernetes_node_taint" "test" {
 }
 
 func testAccKubernetesNodeTaintConfig_removeTaint() string {
-	return fmt.Sprintf(`
-data "kubernetes_nodes" "test" {}
+	return fmt.Sprintf(`data "kubernetes_nodes" "test" {}
 
 resource "kubernetes_node_taint" "test" {
   metadata {

--- a/kubernetes/resource_kubernetes_pod_v1_test.go
+++ b/kubernetes/resource_kubernetes_pod_v1_test.go
@@ -1949,8 +1949,7 @@ func testAccKubernetesPodV1ConfigWithSecurityContextRunAsGroup(podName, imageNam
 }
 
 func testAccKubernetesPodV1ConfigWithSecurityContextSeccompProfile(podName, imageName, seccompProfileType string) string {
-	return fmt.Sprintf(`
-resource "kubernetes_pod_v1" "test" {
+	return fmt.Sprintf(`resource "kubernetes_pod_v1" "test" {
   metadata {
     labels = {
       app = "pod_label"
@@ -1982,8 +1981,7 @@ resource "kubernetes_pod_v1" "test" {
 }
 
 func testAccKubernetesPodV1ConfigWithSecurityContextSeccompProfileLocalhost(podName, imageName string) string {
-	return fmt.Sprintf(`
-resource "kubernetes_pod_v1" "test" {
+	return fmt.Sprintf(`resource "kubernetes_pod_v1" "test" {
   metadata {
     labels = {
       app = "pod_label"

--- a/kubernetes/resource_kubernetes_validating_webhook_configuration_v1_test.go
+++ b/kubernetes/resource_kubernetes_validating_webhook_configuration_v1_test.go
@@ -198,8 +198,7 @@ func testAccCheckKubernetesValidatingWebhookConfigurationV1Exists(n string) reso
 }
 
 func testAccKubernetesValidatingWebhookConfigurationV1Config_basic(name string) string {
-	return fmt.Sprintf(`
-resource "kubernetes_validating_webhook_configuration_v1" "test" {
+	return fmt.Sprintf(`resource "kubernetes_validating_webhook_configuration_v1" "test" {
   metadata {
     name = %q
   }
@@ -235,8 +234,7 @@ resource "kubernetes_validating_webhook_configuration_v1" "test" {
 }
 
 func testAccKubernetesValidatingWebhookConfigurationV1Config_modified(name string) string {
-	return fmt.Sprintf(`
-resource "kubernetes_validating_webhook_configuration_v1" "test" {
+	return fmt.Sprintf(`resource "kubernetes_validating_webhook_configuration_v1" "test" {
   metadata {
     name = %q
   }
@@ -291,8 +289,7 @@ resource "kubernetes_validating_webhook_configuration_v1" "test" {
 }
 
 func testAccKubernetesValidatingWebhookConfigurationV1Config_without_rules(name string) string {
-	return fmt.Sprintf(`
-resource "kubernetes_validating_webhook_configuration_v1" "test" {
+	return fmt.Sprintf(`resource "kubernetes_validating_webhook_configuration_v1" "test" {
   metadata {
     name = %q
   }

--- a/kubernetes/resource_kubernetes_validating_webhook_configuration_v1beta1_test.go
+++ b/kubernetes/resource_kubernetes_validating_webhook_configuration_v1beta1_test.go
@@ -192,8 +192,7 @@ func testAccCheckKubernetesValidatingWebhookConfigurationV1Beta1Exists(n string)
 }
 
 func testAccKubernetesValidatingWebhookConfigurationV1Beta1Config_basic(name string) string {
-	return fmt.Sprintf(`
-resource "kubernetes_validating_webhook_configuration" "test" {
+	return fmt.Sprintf(`resource "kubernetes_validating_webhook_configuration" "test" {
   metadata {
     name = %q
   }
@@ -229,8 +228,7 @@ resource "kubernetes_validating_webhook_configuration" "test" {
 }
 
 func testAccKubernetesValidatingWebhookConfigurationV1Beta1Config_modified(name string) string {
-	return fmt.Sprintf(`
-resource "kubernetes_validating_webhook_configuration" "test" {
+	return fmt.Sprintf(`resource "kubernetes_validating_webhook_configuration" "test" {
   metadata {
     name = %q
   }

--- a/kubernetes/structure_ingress_spec.go
+++ b/kubernetes/structure_ingress_spec.go
@@ -194,12 +194,15 @@ func expandIngressBackend(l []interface{}) *v1beta1.IngressBackend {
 }
 
 func expandIngressTLS(l []interface{}) []v1beta1.IngressTLS {
-	if len(l) == 0 || l[0] == nil {
+	if len(l) == 0 {
 		return nil
 	}
 
 	tlsList := make([]v1beta1.IngressTLS, len(l))
 	for i, t := range l {
+		if t == nil {
+			t = map[string]interface{}{}
+		}
 		in := t.(map[string]interface{})
 		obj := v1beta1.IngressTLS{}
 

--- a/kubernetes/structure_ingress_spec_v1.go
+++ b/kubernetes/structure_ingress_spec_v1.go
@@ -252,12 +252,15 @@ func expandIngressV1Backend(l []interface{}) *networking.IngressBackend {
 }
 
 func expandIngressV1TLS(l []interface{}) []networking.IngressTLS {
-	if len(l) == 0 || l[0] == nil {
+	if len(l) == 0 {
 		return nil
 	}
 
 	tlsList := make([]networking.IngressTLS, len(l))
 	for i, t := range l {
+		if t == nil {
+			t = map[string]interface{}{}
+		}
 		in := t.(map[string]interface{})
 		obj := networking.IngressTLS{}
 

